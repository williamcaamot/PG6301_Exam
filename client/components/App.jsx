import React, {createContext, useEffect, useState} from "react";
import {BrowserRouter, Route, Router, Routes} from "react-router-dom";
import Login from "./login/Login.jsx";
import Header from "./Header.jsx";
import Home from "./home/Home.jsx";
import GoogleLoginCallback from "./login/GoogleLoginCallback.jsx";
import EntraIdLoginCallback from "./login/EntraIdLoginCallback.jsx";
import Chat from "./chat/Chat.jsx";
import Profile from "./profile/Profile.jsx";


export const AppContext = createContext();

export function App() {

    const [user, setUser] = useState(null);

    async function fetchUser() {
        const res = await fetch("/api/v1/login");
        if (res.status === 401) {
            setUser(undefined);

        } else {
            const user = await res.json();
            setUser(user);
            console.log(user)
        }
    }

    useEffect(() => {
        fetchUser();
    }, []);


    return <AppContext.Provider value={{user, setUser}}>
        <BrowserRouter>
        <Header/>
        <Routes>
            <Route path={"/login/callback/entraid"} element={<EntraIdLoginCallback/>}/>
            <Route path={"login/callback/google"} element={<GoogleLoginCallback/>}/>
            <Route path={"/login"} element={<Login/>}/>
            <Route path={"/profile"} element={<Profile/>}/>
            <Route path={"/chat"} element={<Chat/>}/>
            <Route path={"/"} element={<Home/>}/>
        </Routes>


    </BrowserRouter>
    </AppContext.Provider>
}