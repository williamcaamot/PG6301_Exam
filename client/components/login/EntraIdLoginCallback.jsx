import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import ErrorMessage from "../globals/ErrorMessage.jsx";
import { AppContext } from "../App.jsx";

function EntraIdLoginCallback({}) {
  const navigate = useNavigate();
  const [error, setError] = useState();
  const { setUser } = useContext(AppContext);

  const [errorMessage, setErrorMessage] = useState();

  const client_id = "57350161-fcb3-43d0-a7e7-0c2ba9658b8e";

  async function handleCallback() {
    const hash = Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1)),
    );
    let { access_token, error, error_description, state, code } = hash;
    const { token_endpoint } = await fetchJSON(
      "https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration",
    );

    if (state !== window.sessionStorage.getItem("state")) {
      setError("Invalid state");
      return;
    }

    const client_id = "57350161-fcb3-43d0-a7e7-0c2ba9658b8e";

    if (code) {
      const code_verifier = window.sessionStorage.getItem("code_verifier");
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams({
          grant_type: "authorization_code",
          code,
          client_id,
          code_verifier,
        }),
      });
      const json = await res.json();
      access_token = json.access_token;
    }
    if (access_token) {
      const res = await fetch("/api/v1/login/entraid", {
        method: "POST",
        body: JSON.stringify({ access_token: access_token }),
        headers: {
          "content-type": "application/json",
        },
      });
      if (!res.ok) {
        setError(await res.text());
        return;
      }
      const user = await res.json();
      setUser(user);
      navigate("/");
    }
  }

  useEffect(() => {
    handleCallback();
  }, []);

  return (
    <>
      <div className={"pageContentWrapper"}>
        <div className={"innerWrapper"}>
          <h2>Logging in with Entra ID... Please wait</h2>
          <ErrorMessage message={errorMessage} />
        </div>
      </div>
    </>
  );
}

export default EntraIdLoginCallback;

async function fetchJSON(path) {
  const res = await fetch(path);
  if (!res.ok) {
    throw new Error(`Failed to fetch ${path}: ${res.statusText}`);
  }
  return await res.json();
}
