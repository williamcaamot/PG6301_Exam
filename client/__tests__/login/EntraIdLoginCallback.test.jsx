import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { AppContext } from "../../components/App.jsx";
import EntraIdLoginCallback from "../../components/login/EntraIdLoginCallback.jsx";

describe("header component", () => {
  it("display the header with user information", () => {
    const user = {
      _id: "655dd38a4c15f00c20bc3fcf",
      bio: "dawawddawJeg er kul",
      email: "williamcoucheronaamot@gmail.com",
      family_name: "Coucheron-Aamot",
      name: "William",
      nickname: "WILLIAM123wwwww",
      picture:
        "https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c",
    };

    async function setUser() {}

    const component = renderer.create(
      <AppContext.Provider value={{ user, setUser }}>
        <MemoryRouter initialEntries={["/#awdpjdawiojdawoijdawiojdwaiudw"]}>
          <EntraIdLoginCallback />
        </MemoryRouter>
      </AppContext.Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});
