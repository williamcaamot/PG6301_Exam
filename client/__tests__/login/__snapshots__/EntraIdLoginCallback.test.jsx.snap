// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component display the header with user information 1`] = `
<div
  className="pageContentWrapper"
>
  <div
    className="innerWrapper"
  >
    <h2>
      Logging in with Entra ID... Please wait
    </h2>
  </div>
</div>
`;
